scripts/gen-components/program.ts

```diff
@@ -0,0 +1,64 @@
+import * as ts from 'typescript'
+
+import { COMPILER_OPTIONS, TYPE_OUT_FILE, COMPONENT_OUT_FILE } from './constants'
+import { resolveModule } from './resolve'
+
+export const createProgramForModule = (moduleName: string) => {
+  const compilerHost = ts.createCompilerHost(COMPILER_OPTIONS)
+  const resolvedFileName = resolveModule(compilerHost, moduleName, __filename)
+  const program = ts.createProgram({
+    rootNames: [resolvedFileName],
+    options: COMPILER_OPTIONS,
+    host: compilerHost,
+  })
+  const typeChecker = program.getTypeChecker()
+  const sourceFile = program.getSourceFile(resolvedFileName)
+
+  if (!sourceFile) {
+    throw new Error(`Could not get source file for module "${moduleName}"`)
+  }
+
+  const symbol = typeChecker.getSymbolAtLocation(sourceFile)
+
+  if (!symbol) {
+    throw new Error(`Failed to get symbol for module "${moduleName}"`)
+  }
+
+  return {
+    sourceFile,
+    compilerHost,
+    program,
+    typeChecker,
+    symbol,
+  }
+}
+
+export const typeCheckResults = () => {
+  const program = ts.createProgram({
+    rootNames: [TYPE_OUT_FILE, COMPONENT_OUT_FILE],
+    options: COMPILER_OPTIONS,
+  })
+
+  const typeSourceFile = program.getSourceFile(TYPE_OUT_FILE)
+  const componentSourceFile = program.getSourceFile(COMPONENT_OUT_FILE)
+  const diagnostics = [
+    ...ts.getPreEmitDiagnostics(program, typeSourceFile),
+    ...ts.getPreEmitDiagnostics(program, componentSourceFile),
+  ]
+
+  diagnostics.forEach((diagnostic) => {
+    if (diagnostic.file) {
+      const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!)
```

Comment:
maybe I'm missing something, but why is `!` at the end?

Reply:
`start` is possibly `undefined` but `getLineAndCharacterOfPosition` requires a... something... `string` or `number` I think, but we know this exists due to the `diagnostic.file` existing. Taken from this example: https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#a-minimal-compiler

Reply:
Damn this is amazing for typescript to break their own best practices in examples. May be it is old examples. Better approach could be to:

```ts
if (typeof diagnostic.file !== 'undefined' && typeof diagnostic.start !== 'undefined') {}
``` 

to prevent corner cases.

Reply:
That might mean we end up missing some diagnostics. I'll see if there's a way to type guard it and still get them all.